services/vmagent-config-updater/main.go

```diff
@@ -1,56 +1,131 @@
+	jobName                    = newArrayFlag("jobName", "node_exporter", "Scrape job name")
+	targetRequiresK8sAuth      = newArrayFlag("targetRequiresK8sAuth", false, "Defines if target requires K8s auth token")
+	targetsCount               = newArrayFlag("targetsCount", 100, "The number of scrape targets to return from -httpListenAddr. Each target has the same address defined by -targetAddr")
+	targetAddr                 = newArrayFlag("targetAddr", "demo.robustperception.io:9090", "Address with port to use as target address the scrape config returned from -httpListenAddr")
+	scrapeInterval             = newArrayFlag("scrapeInterval", time.Second*5, "The scrape_interval to set at the scrape config returned from -httpListenAddr")
+	scrapeConfigUpdateInterval = newArrayFlag("scrapeConfigUpdateInterval", time.Minute*10, "The -scrapeConfigUpdatePercent scrape targets are updated in the scrape config returned from -httpListenAddr every -scrapeConfigUpdateInterval")
+	scrapeConfigUpdatePercent  = newArrayFlag("scrapeConfigUpdatePercent", 1.0, "The -scrapeConfigUpdatePercent scrape targets are updated in the scrape config returned from -httpListenAddr ever -scrapeConfigUpdateInterval")
+	scrapeConfigMetricRelabel  = newArrayFlag("scrapeConfigMetricRelabel", "", "Path to metric relabel configuration for scrape targets")
 )
 
 func main() {
 	flag.Parse()
 	flag.VisitAll(func(f *flag.Flag) {
-		log.Printf("-%s=%s", f.Name, f.Value)
+		log.Printf("-%s=%s", f.Name, f.Value.String())
 	})
-	c := newConfig(*targetsCount, *scrapeInterval, *targetAddr)
-	var cLock sync.Mutex
-	p := *scrapeConfigUpdatePercent / 100
-	r := rand.New(rand.NewSource(time.Now().UnixNano()))
-	go func() {
-		rev := 0
-		for range time.Tick(*scrapeConfigUpdateInterval) {
-			rev++
-			revStr := fmt.Sprintf("r%d", rev)
-			cLock.Lock()
-			for _, sc := range c.ScrapeConfigs {
-				for _, stc := range sc.StaticConfigs {
-					if r.Float64() >= p {
-						continue
-					}
-					stc.Labels["revision"] = revStr
-				}
-			}
-			cLock.Unlock()
+	uniqueJobs := make(map[string]struct{})
+	for _, job := range jobName.total() {
+		uniqueJobs[job] = struct{}{}
+	}
+
+	log.Printf("creating %d jobs", len(uniqueJobs))
+	targets := make([]*target, len(uniqueJobs))
+	c := &config{
+		ScrapeConfigs: make([]*yaml.Node, len(targets)),
+	}
+	for i := range targets {
+		targets[i] = &target{
+			config: newScrapeConfig(
+				targetsCount.getArg(i),
+				scrapeInterval.getArg(i),
+				targetAddr.getArg(i),
+				labelName.getArg(i),
+				jobName.getArg(i),
+				scrapeConfigMetricRelabel.getArg(i),
+				targetRequiresK8sAuth.getArg(i),
+			),
+			updateInterval: scrapeConfigUpdateInterval.getArg(i),
+			updatePercent:  scrapeConfigUpdatePercent.getArg(i) / 100,
 		}
-	}()
+		go targets[i].run()
+	}
 	rh := func(w http.ResponseWriter, r *http.Request) {
-		cLock.Lock()
+		for i := range targets {
```

Comment:
This code introduces data race. global variable `c` can be update concurrently. It's better to move it into handler.